partiql-parser/src/lalr/lexer.rs

```diff
@@ -1,7 +1,112 @@
+use crate::result::LineAndColumn;
+use itertools::Itertools;
 use logos::{Logos, Span};
+use smallvec::{smallvec, SmallVec};
 use std::cmp::max;
+use std::ops::Range;
 use thiserror::Error;
 
+/// Keeps track of source offsets of newlines for the purposes of later calculating
+/// line and column information
+///
+///
+/// ## Example
+///
+/// ```rust
+/// use partiql_parser::result::LineAndColumn;
+/// use partiql_parser::LineOffsetTracker;
+///
+/// let source = "12345\n789012345\n789012345\n789012345";
+/// let mut tracker = LineOffsetTracker::default();
+/// tracker.record(5..6);
+/// tracker.record(15..16);
+/// tracker.record(25..26);
+///
+/// // We added 3 newlines, so there should be 4 lines of source
+/// assert_eq!(tracker.num_lines(), 4);
+/// assert_eq!(tracker.at(source, 0), LineAndColumn::at(1,1));
+/// assert_eq!(tracker.at(source, 6), LineAndColumn::at(2,1));
+/// assert_eq!(tracker.at(source, 30), LineAndColumn::at(4,5));
+/// ```
+pub struct LineOffsetTracker {
+    line_starts: SmallVec<[usize; 8]>,
+}
+
+impl Default for LineOffsetTracker {
+    fn default() -> Self {
+        LineOffsetTracker {
+            line_starts: smallvec![0], // line 1 starts at offset `0`
+        }
+    }
+}
+
+impl LineOffsetTracker {
+    /// Record a newline at `span` in the source
+    #[inline(always)]
+    pub fn record(&mut self, span: Range<usize>) {
```

Comment:
Does it make sense to only accept `span.end`?

Reply:
From the point of view of `LineOffsetTracker`, yes. 

But, I left it accepting a span to make the Logos lexer definition a little easier to read.
Ideally, `record` will be inlined so it shouldn't matter from a performance perspective.